common/include/verilog_int.h

```diff
@@ -555,29 +556,105 @@ struct vint {
-		// TODO
+		const unsigned shamt =  pos % bw_word;
+		const unsigned lsb_word = pos / bw_word;
+		const stype bmask = stype(1) << shamt;
+		v[lsb_word] &= ~bmask;
+		if (value) {
+			v[lsb_word] |= bmask;
+		}
 	}
 
 	template<bool is_signed>
-	void SetBit(unsigned pos, vint<is_signed, 1u> value) const {
+	void SetBit(unsigned pos, vint<is_signed, 1u> value) {
 		assert(pos < num_bit);
 		SetBit(pos, bool(value));
 	}
 
-	template<unsigned num_bit_src, unsigned pos>
-	bool Slice() const {
-		// TODO
-		return false;
+	template<unsigned pos, unsigned num_bit_slice>
+	vint<false, num_bit_slice> Slice() const {
+		static_assert(pos + num_bit_slice <= num_bit);
+		vint<false, num_bit_slice> sl;
+		constexpr unsigned shamt = pos % bw_word;
+		constexpr unsigned lsb_word = pos / bw_word;
+		constexpr unsigned num_word_slice = detail::num_bit2num_word(num_bit_slice);
+		if constexpr (shamt == 0) {
+			for (unsigned i = 0; i < num_word_slice; ++i) {
+				sl.v[i] = v[i+lsb_word];
+			}
+		} else {
+			for (unsigned i = 0; i < num_word_slice - 1u; ++i) {
+				sl.v[i] = detail::shiftright128(v[i+lsb_word+1], v[i+lsb_word], shamt);
+			}
+			constexpr unsigned required_msb_word = num_word_slice+lsb_word;
+			if constexpr (required_msb_word == num_word) {
+				sl.v[num_word_slice-1] = v[num_word-1] >> shamt;
+			} else {
+				sl.v[num_word_slice-1] = detail::shiftright128(
+					v[required_msb_word],
+					v[required_msb_word-1],
+					shamt
+				);
+			}
+		}
+		sl.ClearUnusedBits();
+		return sl;
+	}
+
+	template<unsigned lsb_pos, unsigned num_bit_slice>
+	void ClearSlice() {
+		constexpr unsigned msb_pos = lsb_pos + num_bit_slice - 1;
+		constexpr unsigned lsb_word_slice = lsb_pos / bw_word;
+		constexpr unsigned msb_word_slice = msb_pos / bw_word;
+		constexpr unsigned unused_lsb_slice = lsb_pos % bw_word;
+		constexpr unsigned unused_msb_slice = bw_word - 1 - (msb_pos % bw_word);
+		constexpr stype lsb_mask_slice = ~(stype(-1) << unused_lsb_slice);
+		constexpr stype msb_mask_slice = ~(stype(-1) >> unused_msb_slice);
+		if constexpr (lsb_word_slice == msb_word_slice) {
+			v[lsb_word_slice] &= lsb_mask_slice | msb_mask_slice;
+		} else {
+			v[lsb_word_slice] &= lsb_mask_slice;
+			for (unsigned i = lsb_word_slice+1; i < msb_word_slice; ++i) {
+				v[i] = 0;
+			}
+			v[msb_word_slice] &= msb_mask_slice;
+		}
 	}
 
-	template<unsigned num_bit_src, unsigned pos>
-	void SetSlice() const {
-		// TODO
+	template<unsigned lsb_pos, unsigned num_bit_slice>
+	void WriteSlice(const vint<false, num_bit_slice>& sl) {
+		constexpr unsigned msb_pos = lsb_pos + num_bit_slice - 1;
+		constexpr unsigned lsb_word_slice = lsb_pos / bw_word;
+		constexpr unsigned msb_word_slice = msb_pos / bw_word;
+		constexpr unsigned num_word_slice = detail::num_bit2num_word(num_bit_slice);
+		constexpr unsigned unused_lsb_slice = lsb_pos % bw_word;
+		if constexpr (unused_lsb_slice == 0) {
+			for (unsigned i = 0; i < num_word_slice; ++i) {
+				v[lsb_word_slice+i] |= sl.v[i];
```

Comment:
The |= seems a little bit violate the name `WriteSlice`. If the caller does not reset the bits, the value is `or` not `write`.
Perhaps make this function private so user cannot call it directly, force them to use SetSlice?

Reply:
Actually I use it elsewhere, let me rename it.